src/ring_theory/non_unital_subsemiring/basic.lean

```diff
@@ -0,0 +1,814 @@
+/-
+Copyright (c) 2022 Jireh Loreaux All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+
+import algebra.module.basic
+import algebra.ring.equiv
+import algebra.ring.prod
+import data.set.finite
+import group_theory.submonoid.centralizer
+import group_theory.submonoid.membership
+import group_theory.subsemigroup.membership
+
+/-!
+# Bundled non-unital subsemirings
+
+We define bundled non-unital subsemirings and some standard constructions:
+`complete_lattice` structure, `subtype` and `inclusion` ring homomorphisms, non-unital subsemiring
+`map`, `comap` and range (`srange`) of a `non_unital_ring_hom` etc.
+-/
+
+open_locale big_operators
+
+universes u v w
+
+variables {R : Type u} {S : Type v} {T : Type w} [non_unital_non_assoc_semiring R]
+  (M : subsemigroup R)
+
+/-- `non_unital_subsemiring_class S R` states that `S` is a type of subsets `s ⊆ R` that
+are both an additive submonoid and also a multiplicative subsemigroup. -/
+class non_unital_subsemiring_class (S : Type*) (R : out_param $ Type u)
+  [non_unital_non_assoc_semiring R] [set_like S R] extends add_submonoid_class S R :=
+(mul_mem : ∀ {s : S} {a b : R}, a ∈ s → b ∈ s → a * b ∈ s)
+
+@[priority 100] -- See note [lower instance priority]
+instance non_unital_subsemiring_class.mul_mem_class (S : Type*) (R : out_param $ Type u)
+  [non_unital_non_assoc_semiring R] [set_like S R] [h : non_unital_subsemiring_class S R] :
+  mul_mem_class S R :=
+{ .. h }
```

Comment:
I think it's for technical reasons to avoid definitional equality checks. Note that on #16134 I had done something similar to what you're suggesting and it caused (seemingly) unrelated timeouts with the `fails_quickly` linter. This is also the approach taken in other parts of the library, so we'll stick with this for now for safety. Hopefully in Lean 4 we won't even have to worry about this.